pageserver/src/tenant.rs

```diff
@@ -350,9 +361,604 @@ impl Drop for TimelineUninitMark {
     }
 }
 
+// We should not blindly overwrite local metadata with remote one.
+// For example, consider the following case:
+//     Checkpoint comes, we update local metadata and start upload task but after that
+//     pageserver crashes. During startup we'll load new metadata, and then reset it
+//     to the state of remote one. But current layermap will have layers from the old
+//     metadata which is inconsistent.
+//     And with current logic it wont disgard them during load because during layermap
+//     load it sees local disk consistent lsn which is ahead of layer lsns.
+//     If we treat remote as source of truth we need to completely sync with it,
+//     i e delete local files which are missing on the remote. This will add extra work,
+//     wal for these layers needs to be reingested for example
+//
+// So the solution is to take remote metadata only when it has greater disk_consistent_lsn
+pub fn merge_local_remote_metadata<'a>(
+    local: Option<&'a TimelineMetadata>,
+    remote: Option<&'a TimelineMetadata>,
+) -> anyhow::Result<(&'a TimelineMetadata, bool)> {
+    match (local, remote) {
+        (None, None) => anyhow::bail!("we should have either local metadata or remote"),
+        (None, Some(remote)) => Ok((remote, false)),
+        (Some(local), None) => Ok((local, true)),
+        (Some(local), Some(remote)) => {
+            // take metadata with highest disk_consistent_lsn.
+            // FIXME Do we need to merge something?
```

Comment:
@LizardWizzard we should be confident about this and remove the FIXME before merging.

AFAICT, the logic below is OK because we always upload full `TimelineMetadata`'s, and the LSN provides an implicit versioning of the file.

What I'm more worried / unclear about is the presence/absence of layer files.

The normal case for "we crashed with upload ops in the queue" is the `Some(local) > Some(remote)` case.
In that case, it's clear that the local layer file set is the authoritative one, because local state always holds all layer files and is always ahead or on par with the remote state.

The "crashed during attach" case is `Some(local) = Some(remote)`. There, the `download_missing` will establish a state where local layer file set is on par with the remote state before we call `reconcile_with_remote`.
So, that's good as well.

What remains is the `Some(local) < Some(remote)` case. I don't really get when this would happen under the operations that we expose.
I guess you can make it happen like so:
-  detaching the tenant but keeping the local FS.
- Then attaching the tenant somewhere else.
- Make some progress there, upload it.
- Then detach there.
- Re-attach here, with local FS state.
I think the attaching marker file that I added prevents that scenario.
And we shouldn't be messing like this with the local FS state anyways.
So, when does this case happen? @LizardWizzard

Reply:
Decision: For `Some(local) < Some(remote)`, we're going to bail out.

Reply:
> The "crashed during attach" case is Some(local) = Some(remote). 

During attach we still write metadata at the end, so it can be `None, Some(remote)`

> I don't really get when this would happen under the operations that we expose.

The repro can be:
1. Some progress for tenant X on pageserver PS1
2. Stop PS1
3. Attach tenant to PS2
4. Make some progress there
5. Start PS1

Bailing looks ok for me as of now. We can remove it with addition of a seesaw test case I described in one of our conversations. 

>  // FIXME Do we need to merge something?

I wonder about gc_horizon, or something else. It should be fine if we have higher gc_horizon, it will be too conservative, and we may still have data needed to create branch past gc horizon locally, but it looks harmless.

@hlinnaka Do you have any ideas regarding potentially dangerous consequences? If no, then lets remove this FIXME

Reply:
For `latest_gc_cutoff_lsn`, I think we need to pick `Max(remote.latest_gc_cutoff_lsn, local.latest_gc_cutoff_lsn)`. That's conservative and safe.

Reply:
> For `latest_gc_cutoff_lsn`, I think we need to pick `Max(remote.latest_gc_cutoff_lsn, local.latest_gc_cutoff_lsn)`. That's conservative and safe.

Thinking about it some more, like with `disk_consisent_lsn`, the `remote.latest_gc_cutoff_lsn > local.latest_gc_cutoff_lsn` case should never happen. We always update the local value first, so if `remote.latest_gc_cutoff_lsn > local.latest_gc_cutoff_lsn`, something else updated the remote value, probably a different pageserver. So I think we could bail out in that case, too.

Reply:
PR: https://github.com/neondatabase/neon/pull/2935